[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98826
* #98730
* __->__ #99132

**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

leslie-fang-intel committed:
Enable the decomposed dequant maxpooling2d loop fusion

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
When lowering of `max_pool2d`, it will check the `num_reads` for input buffer. When `num_reads` is large than 1, `max_pool2d` will fail to do loop fusion with previous ops. In quantization case, the previous op generally will be `decomposed.dequant_per_tensor`. And for the op overload of `decomposed.dequant_per_tensor.tensor`, it has 3 `num_reads`. But 2 of these 3 `num_reads` are scalar tensors as `zero point` and `scale`. In this PR, we will check the `num_of_non_scalar_tensor_read` instead of `num_reads`, which can enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor - max_pool2d`.

**Test Plan**
```
python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
When lowering of `max_pool2d`, it will check the `num_reads` for input buffer. When `num_reads` is large than 1, `max_pool2d` will fail to do loop fusion with previous ops. In quantization case, the previous op generally will be `decomposed.dequant_per_tensor`. And for the op overload of `decomposed.dequant_per_tensor.tensor`, it has 3 `num_reads`. But 2 of these 3 `num_reads` are scalar tensors as `zero point` and `scale`. In this PR, we will check the `num_of_non_scalar_tensor_read` instead of `num_reads`, which can enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
When lowering of `max_pool2d`, it will check the `num_reads` for input buffer. When `num_reads` is large than 1, `max_pool2d` will fail to do loop fusion with previous ops. In quantization case, the previous op generally will be `decomposed.dequant_per_tensor`. And for the op overload of `decomposed.dequant_per_tensor.tensor`, it has 3 `num_reads`. But 2 of these 3 `num_reads` are scalar tensors as `zero point` and `scale`. In this PR, we will check the `num_of_non_scalar_tensor_read` instead of `num_reads`, which can enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented on torch/_inductor/ir.py:
```diff
@@ -3800,6 +3804,32 @@ def num_reads(self):
         return len(read_writes.reads)
 
+    @cache_on_self
+    def check_non_scalar_tensor_num_reads(self):
```
Could you elaborate on why we need this check?

leslie-fang-intel replied:
Thanks for the comment. This check is used to relax the criterion for `StorageBox.realize`. Specifically, when the input is a instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1.
I have also updated the summary of this PR to elaborate the motivation and update the generated kernel before and after this PR.

EikanWang commented on torch/_inductor/ir.py:
```diff
@@ -3800,6 +3804,32 @@ def num_reads(self):
+                if (
+                    len(read.index.free_symbols) == 0
+                    and (read.index.is_Integer)
+                    and (read.index == 0)
```
Does `read.index` imply `read.idex.is_Integer` and `len(read.index.free_symbols) == 0`?

leslie-fang-intel replied:
Thanks and only keep the check of `read.index == 0`.

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
When lowering of `max_pool2d`, it will check the `num_reads` for input buffer. When `num_reads` is large than 1, `max_pool2d` will fail to do loop fusion with previous ops. In quantization case, the previous op generally will be `decomposed.dequant_per_tensor`. And for the op overload of `decomposed.dequant_per_tensor.tensor`, it has 3 `num_reads`. But 2 of these 3 `num_reads` are scalar tensors as `zero point` and `scale`. In this PR, we will check the `num_of_non_scalar_tensor_read` instead of `num_reads`, which can enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
When lowering of `max_pool2d`, it will check the `num_reads` for input buffer. When `num_reads` is large than 1, `max_pool2d` will fail to do loop fusion with previous ops. In quantization case, the previous op generally will be `decomposed.dequant_per_tensor`. And for the op overload of `decomposed.dequant_per_tensor.tensor`, it has 3 `num_reads`. But 2 of these 3 `num_reads` are scalar tensors as `zero point` and `scale`. In this PR, we will check the `num_of_non_scalar_tensor_read` instead of `num_reads`, which can enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
Take the UT pattern of `decomposed.dequant_per_tensor.tensor - max_pool2d` in this PR as example. Note, input is with channel last format.
Before this PR, the generated kernel is:
```
extern "C" void kernel(const unsigned char* in_ptr0,
                       const long* in_ptr1,
                       const float* in_ptr2,
                       float* out_ptr0,
                       float* out_ptr1)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0L); i0<static_cast<long>(3072L); i0+=static_cast<long>(16L))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(i0));
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(in_ptr1[static_cast<long>(0L)]));
                auto tmp5 = at::vec::Vectorized<float>(static_cast<float>(in_ptr2[static_cast<long>(0L)]));
                auto tmp1 = (tmp0);
                auto tmp3 = (tmp2);
                auto tmp4 = tmp1 - tmp3;
                auto tmp6 = tmp4 * tmp5;
                tmp6.store(out_ptr0 + static_cast<long>(i0));
            }
        }
        {
            #pragma omp for  collapse(3)
            for(long i0=static_cast<long>(0L); i0<static_cast<long>(3L); i0+=static_cast<long>(1L))
            {
                for(long i1=static_cast<long>(0L); i1<static_cast<long>(5L); i1+=static_cast<long>(1L))
                {
                    for(long i2=static_cast<long>(0L); i2<static_cast<long>(5L); i2+=static_cast<long>(1L))
                    {
                        for(long i3=static_cast<long>(0L); i3<static_cast<long>(16L); i3+=static_cast<long>(16L))
                        {
                            auto tmp0 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i1)));
                            auto tmp1 = at::vec::Vectorized<int>(static_cast<int>(0));
                            auto tmp2 = to_float_mask(tmp0 >= tmp1);
                            auto tmp3 = at::vec::Vectorized<int>(static_cast<int>(8));
                            auto tmp4 = to_float_mask(tmp0 < tmp3);
                            auto tmp5 = tmp2 & tmp4;
                            auto tmp6 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i2)));
                            auto tmp7 = to_float_mask(tmp6 >= tmp1);
                            auto tmp8 = to_float_mask(tmp6 < tmp3);
                            auto tmp9 = tmp7 & tmp8;
                            auto tmp10 = tmp5 & tmp9;
                            auto tmp11 = [&]
                            {
                                auto tmp12 = at::vec::Vectorized<float>::loadu(out_ptr0 + static_cast<long>((-144L) + i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                return tmp12;
                            }
                            ;
                            auto tmp13 = decltype(tmp11())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp11(), to_float_mask(tmp10));
                            auto tmp14 = at::vec::Vectorized<int>(static_cast<int>(2L*i2));
                            auto tmp15 = to_float_mask(tmp14 >= tmp1);
                            auto tmp16 = to_float_mask(tmp14 < tmp3);
                            auto tmp17 = tmp15 & tmp16;
                            auto tmp18 = tmp5 & tmp17;
                            auto tmp19 = [&]
                            {
                                auto tmp20 = at::vec::Vectorized<float>::loadu(out_ptr0 + static_cast<long>((-128L) + i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                return tmp20;
                            }
                            ;
                            auto tmp21 = decltype(tmp19())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp19(), to_float_mask(tmp18));
                            auto tmp22 = at::vec::maximum(tmp21, tmp13);
                            auto tmp23 = at::vec::Vectorized<int>(static_cast<int>(2L*i1));
                            auto tmp24 = to_float_mask(tmp23 >= tmp1);
                            auto tmp25 = to_float_mask(tmp23 < tmp3);
                            auto tmp26 = tmp24 & tmp25;
                            auto tmp27 = tmp26 & tmp9;
                            auto tmp28 = [&]
                            {
                                auto tmp29 = at::vec::Vectorized<float>::loadu(out_ptr0 + static_cast<long>((-16L) + i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                return tmp29;
                            }
                            ;
                            auto tmp30 = decltype(tmp28())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp28(), to_float_mask(tmp27));
                            auto tmp31 = at::vec::maximum(tmp30, tmp22);
                            auto tmp32 = tmp26 & tmp17;
                            auto tmp33 = [&]
                            {
                                auto tmp34 = at::vec::Vectorized<float>::loadu(out_ptr0 + static_cast<long>(i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                return tmp34;
                            }
                            ;
                            auto tmp35 = decltype(tmp33())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp33(), to_float_mask(tmp32));
                            auto tmp36 = at::vec::maximum(tmp35, tmp31);
                            tmp36.store(out_ptr1 + static_cast<long>(i3 + (16L*i2) + (80L*i1) + (400L*i0)));
                        }
                    }
                }
            }
        }
    }
}
```

After this PR, the generated kernel is:
```
extern "C" void kernel(const unsigned char* in_ptr0,
                       const long* in_ptr1,
                       const float* in_ptr2,
                       float* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for  collapse(3)
            for(long i0=static_cast<long>(0L); i0<static_cast<long>(3L); i0+=static_cast<long>(1L))
            {
                for(long i1=static_cast<long>(0L); i1<static_cast<long>(5L); i1+=static_cast<long>(1L))
                {
                    for(long i2=static_cast<long>(0L); i2<static_cast<long>(5L); i2+=static_cast<long>(1L))
                    {
                        for(long i3=static_cast<long>(0L); i3<static_cast<long>(16L); i3+=static_cast<long>(16L))
                        {
                            auto tmp0 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i1)));
                            auto tmp1 = at::vec::Vectorized<int>(static_cast<int>(0));
                            auto tmp2 = to_float_mask(tmp0 >= tmp1);
                            auto tmp3 = at::vec::Vectorized<int>(static_cast<int>(8));
                            auto tmp4 = to_float_mask(tmp0 < tmp3);
                            auto tmp5 = tmp2 & tmp4;
                            auto tmp6 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i2)));
                            auto tmp7 = to_float_mask(tmp6 >= tmp1);
                            auto tmp8 = to_float_mask(tmp6 < tmp3);
                            auto tmp9 = tmp7 & tmp8;
                            auto tmp10 = tmp5 & tmp9;
                            auto tmp11 = [&]
                            {
                                auto tmp12 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>((-144L) + i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                auto tmp13 = (tmp12);
                                auto tmp14 = at::vec::Vectorized<float>(static_cast<float>(in_ptr1[static_cast<long>(0L)]));
                                auto tmp15 = (tmp14);
                                auto tmp16 = tmp13 - tmp15;
                                auto tmp17 = at::vec::Vectorized<float>(static_cast<float>(in_ptr2[static_cast<long>(0L)]));
                                auto tmp18 = tmp16 * tmp17;
                                return tmp18;
                            }
                            ;
                            auto tmp19 = decltype(tmp11())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp11(), to_float_mask(tmp10));
                            auto tmp20 = at::vec::Vectorized<int>(static_cast<int>(2L*i2));
                            auto tmp21 = to_float_mask(tmp20 >= tmp1);
                            auto tmp22 = to_float_mask(tmp20 < tmp3);
                            auto tmp23 = tmp21 & tmp22;
                            auto tmp24 = tmp5 & tmp23;
                            auto tmp25 = [&]
                            {
                                auto tmp26 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>((-128L) + i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                auto tmp27 = (tmp26);
                                auto tmp28 = at::vec::Vectorized<float>(static_cast<float>(in_ptr1[static_cast<long>(0L)]));
                                auto tmp29 = (tmp28);
                                auto tmp30 = tmp27 - tmp29;
                                auto tmp31 = at::vec::Vectorized<float>(static_cast<float>(in_ptr2[static_cast<long>(0L)]));
                                auto tmp32 = tmp30 * tmp31;
                                return tmp32;
                            }
                            ;
                            auto tmp33 = decltype(tmp25())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp25(), to_float_mask(tmp24));
                            auto tmp34 = at::vec::maximum(tmp33, tmp19);
                            auto tmp35 = at::vec::Vectorized<int>(static_cast<int>(2L*i1));
                            auto tmp36 = to_float_mask(tmp35 >= tmp1);
                            auto tmp37 = to_float_mask(tmp35 < tmp3);
                            auto tmp38 = tmp36 & tmp37;
                            auto tmp39 = tmp38 & tmp9;
                            auto tmp40 = [&]
                            {
                                auto tmp41 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>((-16L) + i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                auto tmp42 = (tmp41);
                                auto tmp43 = at::vec::Vectorized<float>(static_cast<float>(in_ptr1[static_cast<long>(0L)]));
                                auto tmp44 = (tmp43);
                                auto tmp45 = tmp42 - tmp44;
                                auto tmp46 = at::vec::Vectorized<float>(static_cast<float>(in_ptr2[static_cast<long>(0L)]));
                                auto tmp47 = tmp45 * tmp46;
                                return tmp47;
                            }
                            ;
                            auto tmp48 = decltype(tmp40())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp40(), to_float_mask(tmp39));
                            auto tmp49 = at::vec::maximum(tmp48, tmp34);
                            auto tmp50 = tmp38 & tmp23;
                            auto tmp51 = [&]
                            {
                                auto tmp52 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(i3 + (32L*i2) + (256L*i1) + (1024L*i0)));
                                auto tmp53 = (tmp52);
                                auto tmp54 = at::vec::Vectorized<float>(static_cast<float>(in_ptr1[static_cast<long>(0L)]));
                                auto tmp55 = (tmp54);
                                auto tmp56 = tmp53 - tmp55;
                                auto tmp57 = at::vec::Vectorized<float>(static_cast<float>(in_ptr2[static_cast<long>(0L)]));
                                auto tmp58 = tmp56 * tmp57;
                                return tmp58;
                            }
                            ;
                            auto tmp59 = decltype(tmp51())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp51(), to_float_mask(tmp50));
                            auto tmp60 = at::vec::maximum(tmp59, tmp49);
                            tmp60.store(out_ptr0 + static_cast<long>(i3 + (16L*i2) + (80L*i1) + (400L*i0)));
                        }
                    }
                }
            }
        }
    }
}

```

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3800,6 +3804,28 @@ def num_reads(self):
+    @cache_on_self
+    def check_non_scalar_tensor_num_reads(self):
+        data = self.data
+        if isinstance(data, Pointwise):
+            read_writes = ComputedBuffer(
+                name=None,
+                layout=FlexibleLayout(
+                    device=data.get_device(),
+                    dtype=data.get_dtype(),
+                    size=data.get_size(),
+                ),
+                data=data,
+            ).get_read_writes()
```
`Loops` as the parent class of `Pointwise` has `get_reads` function. Can you call that directly?

leslie-fang-intel replied:
Thanks and changed.

jgong5 replied:
What's the problem if we use `Loops.get_reads` here?

leslie-fang-intel replied:
Changed to use `Loops.get_reads` directly. After rebase, I didn't see issues any more.

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3800,6 +3804,28 @@ def num_reads(self):
+                ),
+                data=data,
+            ).get_read_writes()
+            num_of_scalar_tensor = 0
+            for read in read_writes.reads:
+                if read.index == 0:
+                    num_of_scalar_tensor += 1
+            return (len(read_writes.reads) - num_of_scalar_tensor) > 1
```
is it equivalent to `any(read.index != 0 for read in read_writes.reads)`?

leslie-fang-intel replied:
Thanks for the suggestions. I have simplified this implementation. Since we need to count the `non scalar tensor` number and compare with 1, instead of use `any`, we can use `sum` in this case.

jgong5 reviewed (commented):
Can you share the perf numbers the kernel before and after the change? We'd better also check the performance impact on other models...

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
> Can you share the perf numbers the kernel before and after the change? 

I write a benchmark https://gist.github.com/leslie-fang-intel/e0cf371b4073fc6dd7941637bc4452ae to test the performance of this pattern. To simulate the case in RN50, we use input size [116, 64, 112, 112] and use the `channel_last` format as input. The benchmark result on CPX single socket is:

- Before this patch: 813.344 ms
- After this patch: 423.656 ms

> We'd better also check the performance impact on other models...

I will prepare a test branch and ask help to compare the performance with latest weekly. Will post the performance data when done.

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
Here is the test branch https://github.com/leslie-fang-intel/pytorch/tree/leslie/test_inductor_performance_after_relax_relalize_hint_check based on PyTorch Commit https://github.com/leslie-fang-intel/pytorch/commit/319a45d and apply this PR. Su Dong helps to compare the performance of above branches. Here is the details of the summary and no new failures and perf regression found.
Here is the [test branch](https://github.com/leslie-fang-intel/pytorch/tree/leslie/test_inductor_performance_after_relax_relalize_hint_check) based on [PyTorch Commit](https://github.com/leslie-fang-intel/pytorch/commit/319a45d) and apply this PR.  [ESI-SYD](https://github.com/ESI-SYD)   helps to compare the performance of above branches and no new failures and perf regression found. Here is the details of the summary:

![image](https://user-images.githubusercontent.com/53841472/233908476-102ec2ed-bde7-42b1-8aa2-6d94414fe0e5.png)

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3929,21 +3929,12 @@ def num_reads(self):
+        # Skip the check for non Pointwise instances
+        return (
+            (sum(read.index != 0 for read in self.data.get_reads()) > 1)
+            if isinstance(self.data, Pointwise)
```
If you only care about `Pointwise` here, shall we rename the function name to reflect this?

leslie-fang-intel replied:
Thanks for the suggestions and changed.

jgong5 reviewed (approved):
LGTM.

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jansel, could you please kindly help to take a look of this PR?

jansel commented on torch/_inductor/ir.py:
```diff
@@ -3923,6 +3927,15 @@ def num_reads(self):
         return len(read_writes.reads)
 
+    @cache_on_self
+    def is_Pointwise_non_scalar_tensor_num_reads_larger_than_one(self):
```
```suggestion
    def is_pointwise_non_scalar_tensor_num_reads_larger_than_one(self):
```

leslie-fang-intel replied:
Thanks and changed.

leslie-fang-intel committed:
Update on "[Inductor][Quant]Enable the decomposed dequant maxpooling2d loop fusion"


**Summary**
Lowering of [`max_pool2d` ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2732) will check the `num_reads` of input `StorageBox.data`. When num of reads is larger than 1, input of `StorageBox` will invoke `realize` and break the loop fusion with previous node. The previous node could be `decomposed.dequant_per_tensor.tensor` in quantization use case. For `decomposed.dequant_per_tensor.tensor`, it has 3 num of reads. But 2 of these 3 num of reads are scalar tensors as `zero point` and `scale`. In this PR, we try to relax the criterion for `StorageBox.realize`. Specifically, when the input is an instance of `Pointwise`, we will also check the number of non scalar tensor's read, and only invoke `StorageBox.realize` when the number of non scalar tensor's read is also larger than 1. It helps enable the loop fusion and vec code gen of pattern `decomposed.dequant_per_tensor.tensor - max_pool2d`.

**Test Plan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_maxpool2d_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
